### PR TITLE
Clean up SortExec creation and add doc comments

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

# Rationale for this change

While reviewing https://github.com/apache/arrow-datafusion/pull/5881 I had to look up what exactly "new_with_partitioning" did and what all the arguments meant and found the whole thing quite obtuse. Let's make this clearer.

# What changes are included in this PR?

1. Add doc comments better explaining what SortExec does and what `preserve_partitioning` does
1. Add `SortExec::new()` which is infallable (try_new always returned true)
2. Add `SortExec::with_preserve_partitioning` to set the preserve partitioning flag
2. Add `SortExec::with_fetch` to set the fetch
4. Mark `SortExec::try_new()` and `SortExec::new_with_partitioning` as deprecated

# Are these changes tested?

Yes, by clippy and existing tests

# Are there any user-facing changes?

The old APIs are deprecated, but they will still work

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/sort_enforcement.rs [hunk 1]
```diff
@@ -1220,12 +1220,10 @@ mod tests {
...
         let repartition_exec = repartition_exec(spm);
-        let sort2 = Arc::new(SortExec::new_with_partitioning(
-            sort_exprs.clone(),
-            repartition_exec,
-            true,
-            None,
-        )) as _;
+        let sort2 = Arc::new(
+            SortExec::new(sort_exprs.clone(), repartition_exec, None)
+                .with_preserve_partitioning(true),
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/sort_enforcement.rs:1225]: I think this makes it clearer that the sort is being constructed to  preserve the input partitioning

**Review by tustvold — APPROVED** [review]

**Comment by alamb** [comment]
WHile looking at the code again, I realized that `fetch` also was almost always `None` so rather than having to provide a `None` at all creation sites, I added a `with_fetch` function -- since I was changing the signature of `new()` anyways I figured I might as well do that as well in the same PR

File: datafusion/core/src/physical_plan/sorts/sort.rs [hunk 2]
```diff
@@ -628,42 +631,74 @@ pub struct SortExec {
...
+    ///
+    /// If `preserve_partitioning` is false, sorts and merges all
+    /// input partitions producing a single, sorted partition.
+    pub fn with_preserve_partitioning(mut self, preserve_partitioning: bool) -> Self {
+        self.preserve_partitioning = preserve_partitioning;
+        self
+    }
+
+    /// Whether this `SortExec` preserves partitioning of the children
+    pub fn with_fetch(mut self, fetch: Option<usize>) -> Self {
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/sorts/sort.rs:699]: I wonder if limit might be a more standard name for this?

Comment (alamb) [reply datafusion/core/src/physical_plan/sorts/sort.rs:699]: I think the term comes from Spark (some earlier versions of the code had the word Limit). In any event I think `fetch` is at least consistent with the rest of the DataFusion codebase, such as in the implementation of sort:

https://github.com/apache/arrow-datafusion/blob/d9cffa6b18e0796dc718cec3ab94bc0119da747a/datafusion/core/src/physical_plan/sorts/sort.rs#L335

**Pull request merged**